bokeh/models/ranges.py

```diff
@@ -119,6 +142,16 @@ class FactorRange(Range):
     this categorical range.
     """)
 
+    bound_lower = Either(String, Int, help="""
+    The minimum value that the range is allowed to go to - typically used to prevent
+    the user from panning/zooming/etc away from the data.
+    """)
+
+    bound_upper = Either(String, Int, help="""
+    The max value that the range is allowed to go to - typically used to prevent
+    the user from panning/zooming/etc away from the data.
+    """)
```

Comment:
This could work, except that now users are always required to supply both bounds what if they only want a bound below? Could potentially use ellipsis for that: `bounds = (..., 10)` it's slightly unusual but not terribly so. I'd keep to mathematical convention and have it always be `(min, max)`.  There is also some slight discrepancy in that in mathematics `(0,10)` and `[0,10]` are not technically the same, but we would treat all intervals as closed (the latter).

Another nice thing: I was just thinking today that it would be nice to be able to easily hook up an axis bounds to a data range bounds and this would make that simple. 

Last comment, `Either(Tuple(Float, Float), Tuple(Datetime, Datetime), Tuple(Int, Int), Auto, None, default=Auto)` is alot, we might want to make a new property or maybe even a model for `Bounds`.